qsimcirq/qsim_simulator.py

```diff
@@ -303,8 +307,12 @@ def _sample_measure_results(
             ValueError: If there are multiple MeasurementGates with the same key,
                 or if repetitions is negative.
         """
-        if not isinstance(program, qsimc.QSimCircuit):
-            program = qsimc.QSimCircuit(program, device=program.device)
+
+        # Add noise to the circuit if a noise model was provided.
+        program = qsimc.QSimCircuit(
+            self.noise.noisy_moments(program, sorted(program.all_qubits())),
```

Comment:
Here and in three places below, `program.all_qubits()` can be called only once. It seems there is also some additional overhead if no noise model is passed to the simulator.